### ISSUE TITLE
Callback errors should be properly handled

### ISSUE THREAD

**Issue by GroteGnoom** [igraph/igraph#1791]
When a callback is called, and an error is returned, the kind of error should not just be ignored. 

For example, in `igraph_community_leading_eigenvector()`:
```
if (callback) {
    [...]
    ret = callback([...]);
    if (ret) {
        break;
    }
}
```
This will work fine in most cases, but when an unexpected error is returned, for example an `IGRAPH_ENOMEM`, the function will continue running, but with unexpected results. This shouldn't happen, and the error should be propagated instead.

If I understand its usage correctly, `IGRAPH_INTERRUPTED` can be used specifically when the error should be ignored. We should add that to the error descriptions in `igraph_error.h`. `IGRAPH_CHECK` can maybe be changed to handle this case correctly, so `IGRAPH_CHECK` can also be used for callbacks.

----

List of functions using callbacks:

 - [x] igraph_community_leading_eigenvector
 - [ ] igraph_cliques_callback
 - [ ] igraph_maximal_cliques_callback
 - [x] igraph_motifs_randesu_callback
 - [x] igraph_isomorphic_function_vf2
 - [x] igraph_subisomorphic_function_vf2
 - [x] igraph_bfs
 - [x] igraph_dfs
 - (please extend list as needed)

**Comment by ntamas** [comment]
> We should add that to the error descriptions in `igraph_error.h`. 

It is already there (error code = 13, description = `Interrupted`). And I think it's okay to let this error code propagate through `IGRAPH_CHECK()` just like anything else; callers that wish to support interruption in a different way should catch the error code themselves. We can add an alternative `IGRAPH_CHECK_INTERRUPTED(...)` macro or something like that to have a macro for ourselves where we handle `IGRAPH_INTERRUPTED`.

**Comment by GroteGnoom** [comment]
> It is already there (error code = 13, description = `Interrupted`). 
Yes, but it's not in the documentation in the header file.

**Comment by szhorvat** [comment]
Indeed, the `IGRAPH_INTERRUPTED` error code is not really special in any way. My understanding is that it is handled the same way as any other error code. It should result in an error being reported about the calculation being interrupted.  The difference is that the description informs that user that the abort happened because it was requested, not because of a problem.

In high-level interfaces, the result may appear to be different because the host language will also handle the interrupt request. For example, the Mathematica interface does not show an error message when the computation is interrupted. This is because the code that would print the error also gets interrupted, so in the end we get the desired behaviour (the standard behaviour for interruptions).

Maybe this is how we get KeyboardInterrupt in Python yoo? I don'y know.

**Comment by szhorvat** [comment]
I was interrupted, so continuing now.

The difficulty with returning errors from callback functions is the following:

Currently, most callback functions have a true/false return value which controls whether the search should continue or stop. There is no error.

There are several which incorrectly use `IGRAPH_CHECK`, despite not really returning an error code.

If we were to make them return an error code, then we have a problem: how do we indicate that the search should stop, but there is no error? This is not `IGRAPH_INTERRUPTED`. That is for interactive interruption, and is really like an error for all purposes.  

There is currently one place in igraph where this is implemented: in the maximal clique callback function. To make it work, I introduced a special "error code", `IGRAPH_STOP`, which signals that the search should stop, but there is no error. This is not the same as `IGRAPH_INTERRUPTED`, which should be propagated up to the caller. `IGRAPH_STOP` should specifically be handled, and _not_ passed on. This is a workable solution for other callback functions too, but it always felt like a hack to me.

What would be a better solution? We could separate the error information (return code) from the stop/continue signal (an output parameter). This will involve a lot of API breaking and will complicate things.

@ntamas @GroteGnoom What do you think?  Should we use `IGRAPH_STOP` as a special error (requiring special handling) that signals "there is no error but please stop", or should we use an output parameter for requesting a stop?

**Comment by szhorvat** [comment]
Related: https://github.com/igraph/igraph/pull/1192

**Comment by ntamas** [comment]
I'm adding a bit more context here to clarify the distinction between `IGRAPH_INTERRUPTED` and `IGRAPH_STOP` as it may not be immediately obvious unless you are familiar with how higher-level interfaces work.

`IGRAPH_STOP` is actually used for what started the original discussion between us on Slack: it is a way to stop a callback-based search function prematurely, but _normally_. For instance, if you are looking for the first five maximal cliques but you don't care about the rest, you fill a list with them in the callback until the size of the list reaches 5, then return `IGRAPH_STOP`. The caller swallows `IGRAPH_STOP`, stops the search, cleans up after itself and returns `IGRAPH_SUCCESS` to _its_ own caller.

`IGRAPH_INTERRUPTED` is used for interrupting a long-running calculation from higher level interfaces, based on the user's request. The way it works in the Python interface is that the `IGRAPH_ALLOW_INTERRUPTION()` method regularly calls into Python and invokes `PyErr_CheckSignals()`, giving Python an opportunity to raise a `KeyboardInterrupt` exception if Ctrl-C was pressed by the user. If this happened, then we return `IGRAPH_INTERRUPTED` to igraph, which in turn terminates the calculation, cleans up after itself, and then _re-raises the exception_ so the user can see (and maybe handle) `KeyboardInterrupt`.

**Comment by ntamas** [comment]
Personally, I think the system we have now is satisfactory for all use-cases, we just need to refactor the callbacks to return an error code and make use of `IGRAPH_STOP` and `IGRAPH_INTERRUPTED` consistently. We need to instruct users writing callback functions to return `IGRAPH_STOP` (and _not_ `IGRAPH_INTERRUPTED`) to indicate that they do not need more iterations of the algorithm, and we need to have a special `IGRAPH_CHECK_CALLBACK(...)` macro that is like `IGRAPH_CHECK()` but handles `IGRAPH_STOP` gracefully.

I'd rather not put this straight into `IGRAPH_CHECK()` because some bugs would then be harder to track down; I'm thinking about cases when we call `IGRAPH_CHECK(some_buggy_function())` and `some_buggy_function()` accidentally returns `IGRAPH_STOP`. If we swallowed `IGRAPH_STOP` at that point, we would only see that the function calling `some_buggy_function()` returned with `IGRAPH_SUCCESS` without doing its job.

**Comment by szhorvat** [comment]
I added a list of functions using callbacks.

In addition to callback functions used to decide whether to continue the search, we also have things like:

 - `igraph_isocompat_t`. These are unlikely to need to generate errors, but nor is it ruled out that they could. I would cautiously say that these don't need to be able to handle errors.
 - `igraph_bfshandler_t` and `igraph_dfshandler_t`. I can very well imagine that these could generate errors. Imagine that we record information (= allocate memory) during the search.
 - ???

**Comment by ntamas** [comment]
Most of the callback functions now return an `igraph_error_t`, with the notable exception of the clique-related functions. I think these require modifications to the vendored Cliquer as well. I'll take a closer look at these tomorrow.

**Comment by szhorvat** [comment]
I think it's alright to make modifications to Cliquer. The Cliquer author said that he is not planning further releases. A small fix is already made within igraph. We might as well make more changes.

**Comment by ntamas** [comment]
Cliquer functions fixed in 392d943df . It turned out to be the classic case with the banana and the monkey and the jungle, but it all seems sorted out now.
